[FEATURE] Expressions and algorithms for line substrings

nyalldawson opened the pull request:
Adds API, expression functions, and processing algorithms for calculating line substrings (i.e. a section of a line/curve between a given start and end distance from the beginning of the line).

If z or m values are present, the output z and m will be interpolated using the existing vertices' z or m values.

Handles curved geometries perfectly without segmentization.

Here's a quick screenie using the new function in a geometry generator:

![image](https://user-images.githubusercontent.com/1829991/43941761-af46c4b0-9cb8-11e8-9481-d3b272d4e577.png)

nyalldawson committed:
Add QgsGeometryUtils method for interpolating point between two points, with z/m handling

nyalldawson committed:
Add QgsGeometryUtils method to interpolate a point on an arc given a distance

nyalldawson committed:
[FEATURE] New geometry API call to return a curve substring

Returns a new curve representing a substring of a curve, from
a start distance and end distance.

If z or m values are present, the output z and m will be interpolated using
the existing vertices' z or m values.

Handles curved geometries without loss.

nyalldawson committed:
[FEATURE] line_substring function

Returns the portion of a line (or curve) geometry which falls
between the specified start and end distances (measured from the
beginning of the line). Z and M values are linearly interpolated
from existing values.

Fixes #15611

nyalldawson committed:
[FEATURE][processing] Line substring algorithm

This algorithm returns the portion of a line (or curve) which falls
between the specified start and end distances (measured from the
beginning of the line).

Z and M values are linearly interpolated from existing values.

nyalldawson committed:
[FEATURE][processing] Interpolate point on line algorithm

This algorithm creates a point geometry interpolated at a
set distance along line (or polygon boundary) geometries.

NathanW2 commented:
Holy heck! Yes please

On Fri, Aug 10, 2018 at 4:16 PM Nyall Dawson <notifications@github.com>
wrote:

> Adds API, expression functions, and processing algorithms for calculating
> line substrings (i.e. a section of a line/curve between a given start and
> end distance from the beginning of the line).
>
> If z or m values are present, the output z and m will be interpolated
> using the existing vertices' z or m values.
>
> Handles curved geometries perfectly without segmentization.
>
> Here's a quick screenie using the new function in a geometry generator:
>
> [image: image]
> <https://user-images.githubusercontent.com/1829991/43941761-af46c4b0-9cb8-11e8-9481-d3b272d4e577.png>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/qgis/QGIS/pull/7587
> Commit Summary
>
>    - Add QgsGeometryUtils method for interpolating point between two
>    points, with z/m handling
>    - Add QgsGeometryUtils method to interpolate a point on an arc given a
>    distance
>    - [FEATURE] New geometry API call to return a curve substring
>    - [FEATURE] line_substring function
>    - [FEATURE][processing] Line substring algorithm
>    - [FEATURE][processing] Interpolate point on line algorithm
>
> File Changes
>
>    - *M* python/core/auto_generated/geometry/qgscircularstring.sip.in
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-0> (2)
>    - *M* python/core/auto_generated/geometry/qgscompoundcurve.sip.in
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-1> (2)
>    - *M* python/core/auto_generated/geometry/qgscurve.sip.in
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-2> (14)
>    - *M* python/core/auto_generated/geometry/qgsgeometryutils.sip.in
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-3> (18)
>    - *M* python/core/auto_generated/geometry/qgslinestring.sip.in
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-4> (2)
>    - *A* python/plugins/processing/tests/testdata/airports.gfs
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-5> (43)
>    - *M*
>    python/plugins/processing/tests/testdata/custom/circular_strings.gpkg
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-6> (0)
>    - *A*
>    python/plugins/processing/tests/testdata/expected/airportsvoronoidiagram.gfs
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-7> (43)
>    - *A*
>    python/plugins/processing/tests/testdata/expected/interpolate_point_lines.gml
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-8> (48)
>    - *A*
>    python/plugins/processing/tests/testdata/expected/interpolate_point_lines.xsd
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-9> (23)
>    - *A*
>    python/plugins/processing/tests/testdata/expected/interpolate_point_polys.gml
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-10> (61)
>    - *A*
>    python/plugins/processing/tests/testdata/expected/interpolate_point_polys.xsd
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-11> (43)
>    - *A*
>    python/plugins/processing/tests/testdata/expected/line_substring.gml
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-12> (48)
>    - *A*
>    python/plugins/processing/tests/testdata/expected/line_substring.xsd
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-13> (23)
>    - *M*
>    python/plugins/processing/tests/testdata/qgis_algorithm_tests.yaml
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-14> (39)
>    - *A* resources/function_help/json/line_substring
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-15> (9)
>    - *M* src/analysis/CMakeLists.txt
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-16> (2)
>    - *A* src/analysis/processing/qgsalgorithminterpolatepoint.cpp
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-17> (127)
>    - *A* src/analysis/processing/qgsalgorithminterpolatepoint.h
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-18> (69)
>    - *A* src/analysis/processing/qgsalgorithmlinesubstring.cpp
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-19> (163)
>    - *A* src/analysis/processing/qgsalgorithmlinesubstring.h
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-20> (71)
>    - *M* src/analysis/processing/qgsnativealgorithms.cpp
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-21> (4)
>    - *M* src/core/expression/qgsexpressionfunction.cpp
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-22> (37)
>    - *M* src/core/geometry/qgscircularstring.cpp
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-23> (127)
>    - *M* src/core/geometry/qgscircularstring.h
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-24> (1)
>    - *M* src/core/geometry/qgscompoundcurve.cpp
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-25> (31)
>    - *M* src/core/geometry/qgscompoundcurve.h
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-26> (1)
>    - *M* src/core/geometry/qgscurve.h
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-27> (14)
>    - *M* src/core/geometry/qgsgeometryutils.cpp
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-28> (56)
>    - *M* src/core/geometry/qgsgeometryutils.h
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-29> (32)
>    - *M* src/core/geometry/qgslinestring.cpp
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-30> (90)
>    - *M* src/core/geometry/qgslinestring.h
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-31> (1)
>    - *M* tests/src/core/testqgsexpression.cpp
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-32> (4)
>    - *M* tests/src/core/testqgsgeometry.cpp
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-33> (120)
>    - *M* tests/src/core/testqgsgeometryutils.cpp
>    <https://github.com/qgis/QGIS/pull/7587/files#diff-34> (136)
>
> Patch Links:
>
>    - https://github.com/qgis/QGIS/pull/7587.patch
>    - https://github.com/qgis/QGIS/pull/7587.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/7587>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3GQo8jlPBIX3lVFVXankXqnLhOo2ks5uPSUwgaJpZM4V3pY4>
> .
>

andreasneumann commented:
Very nice - especially that it keeps curve geometries! Thanks!

nyalldawson committed:
[FEATURE] Use native interpolate point method instead of GEOS method

Because:
- Exactly follows curves and doesn't require segmentizing input geometry
- Also interpolates z/m values if they are present in input geometry
- Is faster

nyalldawson commented:
Now comes with a native interpolate point method too, which also respects curves exactly and interpolates Z/M values if present

nyalldawson commented:
Anyone want to review this for me?

pblottiere commented:
> Anyone want to review this for me?

I'm taking a look.

pblottiere commented on src/analysis/processing/qgsalgorithmlinesubstring.cpp:
```diff
@@ -0,0 +1,163 @@
+      {
+        if ( collection->numGeometries() > 0 )
+        {
+          curve = qgsgeometry_cast< const QgsCurve * >( collection->geometryN( 0 ) );
```
What do you think about adding a note in the help string for this particular case?

nyalldawson replied:
Actually, I should handle this case properly - I'll upgrade the algorithm to do that.

nyalldawson replied:
Looking into this, it's non trivial to implement and I suspect the results will never be as intended - I'll just make a note of the current behaviour in the docs instead.

pblottiere commented on src/core/geometry/qgslinestring.cpp:
```diff
@@ -711,6 +711,158 @@ QgsLineString *QgsLineString::reversed() const
   return copy;
 }
 
+QgsPoint *QgsLineString::interpolatePoint( const double distance ) const
```
> const double distance

Nice :). Is the *const correctness* a goal? I mean, do we want to check it during reviews?

nyalldawson replied:
I think it's good practice for new code, but there's SO many existing instances where const is missing that I don't think it's worth including in reviews.

What we really need is for clang-tidy or clazy to implement an automatic fixit for this situation so that we can auto add const to all existing code (when it's valid), and only then move to using const as common practice.

pblottiere commented:
I can't fully understand all the intricacies of interpolation algorithms, but the code looks good! Moreover, considering the amount of unit tests added in this PR, I think we may be confident :).

Thanks Nyall!

nyalldawson commented:
Thanks for the review @pblottiere  - it's much appreciated!

nyalldawson committed:
Expands docs
